### PR TITLE
Save iEta/iPhi & iX/iY indices of ECAL crystals in nanoAOD

### PR DESCRIPTION
**Pull request by swagata87**
#### PR description:
This PR adds some new variables to nano, they are: iEta,iPhi (for barrel) and iX,iY(for endcaps). 
The variables iX, iY will be useful to veto the EE+ leak region at analysis level. 
However, this is not the only use case foreseen, as these variables are useful for various studies. 

#### PR validation:
Workflow `2500.601` ran fine and nano DQM shows that these variables are properly filled. 
Screen Shot 2022-12-15 at 14 18 40 Screen Shot 2022-12-15 at 14 18 50

Screen Shot 2022-12-15 at 14 19 01  Screen Shot 2022-12-15 at 14 19 12

**This PR needs a backport to 12_6_X for the renano of 2022 data.** 

Thanks to @Sam-Harper for useful discussions. In the interest of time, for now I've made the minimal change needed, only in  `PhysicsTools/NanoAOD` package, to make it work. But after discussions with Sam it seems that it would be useful to add these important variables directly to `DataFormats/EgammaReco/interface/SuperCluster.h`. In that case, accessing those variables from nano cff files will be easier, and there won't be any need of a new producer. And it would be more user-friendly for analysers whose setup is based on miniAOD. However, this approach needs some more time to converge. We will revisit it in near future. But for the immediate rush of 12_6_X renano, the current approach is sufficient.

### CODE REVIEW

**Comment by swagata87** [comment]
type egamma

**Comment by swertz** [comment]
enable nano

**Comment by swertz** [comment]
please test

**Comment by swertz** [comment]
urgent

**Comment by perrotta** [comment]
> urgent

HI @swertz. Could you please justify the urgency here? Do you want to add (the backport of) this PR into a possible new 12_6_X patch release right now, since CMSSW_12_6_0 was already buit yesterday with whatever was requested for it during the orp meeting less than two days ago? The place where to request the merge of the PRs and the build of new release is the orp meeting, or in messages sent to the release managers in time for it, definitely not a sparse info in some github thread a few hours after the release was built...

**Comment by swertz** [comment]
Hi @perrotta ,

> Do you want to add (the backport of) this PR into a possible new 12_6_X patch release right now, since CMSSW_12_6_0 was already buit yesterday

Indeed that is the goal.

> The place where to request the merge of the PRs and the build of new release is the orp meeting, or in messages sent to the release managers in time for it, definitely not a sparse info in some github thread a few hours after the release was built...

Apologies. We were only made aware yesterday at the PPD coordination that this would be important to include in the 12_6_X nanoV11 production, so too late for the ORP meeting. I actually thought PPD had discussed it with you and that you were already aware, which is why we didn't you contact you explicitly about it.

**Comment by perrotta** [comment]
Ok, thank you @swertz 
Fine, we will prepare a 12_6_0_patch1 release as soon as the backport of this PR is ready for it.
I just wanted to remark the fact that the plans for the releases must be brought to the ORP meeting: otherwise we can just skip that meeting, and rely on what we read in the github threads instead...

**Comment by swagata87** [comment]
Yes sorry it's partly my oversight as well. 
The recommendation that egamma will provide to analysers for EE+ leak is still being finalised within EGM. 
In the rush of several parallel activities, I had missed that these variables that analysis teams would need to implement our upcoming recommendation is missing in nanoAOD. 
Apologies.

**Comment by swertz** [comment]
+1

New variables look reasonable, I assume the peak at zero is because the variables only make sense for one of barrel or endcap: 
image

**Comment by swagata87** [comment]
Zero is actually not an acceptable value in any of the cases, 
iX and iY run from 1 to 100.
iEta runs from -85 to +85, with no crystal at iEta=0
iPhi runs from 1 to 360. 
The zeros are appearing in iX/iY (iEta/iPhi) plots only when the e/gammas are in barrel (endcaps). 
But initializing with -99 works too. In that case `uint8`->`int8` change will be needed for iY and iY. 
Let me know if -99 would be a better initialisation value. I can change it.

**Comment by swertz** [comment]
I don't have a strong opinion on the default value, but it's true that at least indicating in the docstring that 0 is invalid would be good...

**Comment by perrotta** [comment]
If zero is an unacceptable value in all cases, it can remains as such, provided that it gets documented for the user, as suggested by Sebastien.
What about my suggestion to use only two variables, instead of four, with a different meaning in the barrel and in the endcap. Is it feasible, or it has any drawback, instead?

**Comment by swagata87** [comment]
> What about my suggestion to use only two variables, instead of four, with a different meaning in the barrel and in the endcap. Is it feasible, or it has any drawback, instead?

this should be possible. Let me check.

**Comment by swertz** [comment]
> What about my suggestion to use only two variables, instead of four, with a different meaning in the barrel and in the endcap. Is it feasible, or it has any drawback, instead?

How would you then know whether the SC is the in the barrel or endcap, to apply different cuts? Would it work using the eta of the SC (`Electron_eta+Electron_deltaEtaSC`)? Or perhaps by adding e.g. 1000 to the endcap variables (=> `iX = (iVar1 > 1000) ? (iVar1 % 1000) : 0`, but then it must be an `int` not an `int8`)?

In any case, I don't think size is a problem here... Eight int8's of which half are always zero and compressed away are not a big deal (0.01 kb/event); I doubt it will be very different with four int's instead.

**Comment by swagata87** [comment]
If we go for 2 ints, then they can be named like, `seediEtaOriX` and `seediPhiOriY`. Users will check if the electron/photon is in barrel or in endcap (using supercluster eta cut, as mentioned by @swertz). If the e/gamma supercluster is in barrel then the seed crystal is also in barrel. In that case `seediEtaOriX` and `seediPhiOriY` will essentially be `seediEta` and `seediPhi`. Similar argument for endcap case.

From users point of view, having 4 ints might be more clear and straight forward to understand. So if size is not an issue, we can keep it. But @perrotta let us know your opinion. 

If we are staying with 4 vars, then I will add another commit to update the doc-string to indicate the acceptable range of the new variables.

**Comment by perrotta** [comment]
Thank you @swagata87 and @swertz : I just wanted to know whether the idea was feasible or not. If it brings complication, and due to the limited time we have in front to prepare a possible update, I am in favour of merging the pr as it is in the master, so that you can go ahead with the backport.

Maybe this is something that you can take into account for possibile future developments. I understand that the size of those field is small, but still it looks like a wast of resources (even if probably negligible) to fill and store  objects that you know since the beginning that must be empty. I imagine that a link to the original clusters should be easily provided, and therefore also to its barrel/endcap position in the detector: otherwise, how do you associate these values to the corresponding cluster? But let postpone this discussion to some possible future development.

In summary, I'm going to merge this PR as it is, then.

**Comment by perrotta** [comment]
I'm reading this after having sent the previos message of mine here above.

> If we go for 2 ints, then they can be named like, `seediEtaOriX` and `seediPhiOriY`. Users will check if the electron/photon is in barrel or in endcap (using supercluster eta cut, as mentioned by @swertz). If the e/gamma supercluster is in barrel then the seed crystal is also in barrel. In that case `seediEtaOriX` and `seediPhiOriY` will essentially be `seediEta` and `seediPhi`. Similar argument for endcap case.


This is indeed what I had in mind

> From users point of view, having 4 ints might be more clear and straight forward to understand. So if size is not an issue, we can keep it. But @perrotta let us know your opinion.

I'd follow your directions here: I just made some suggestion, but you must decide what is better for the end user.

> If we are staying with 4 vars, then I will add another commit to update the doc-string to indicate the acceptable range of the new variables.

Ah, ok. I can sit down and wait then: I'd suggest to try to speed up the update, so that we can have the backport merged and the patch release ready before the break.

**Comment by swagata87** [comment]
`seediEtaOriX` and `seediPhiOriY` should work for end-users. EGM will provide instructions twiki for analysers anyway. So it will be clear to them. 

Give me some time to implement these in a final commit.

**Comment by swertz** [comment]
please test

**Comment by swagata87** [comment]
New vars:
Screen Shot 2022-12-16 at 11 23 15 Screen Shot 2022-12-16 at 11 23 27

**Comment by swertz** [comment]
please test

**Comment by swertz** [comment]
+1

Differences in 11634.7 are spurious.

**Comment by vlimant** [comment]
@swagata87 if you can please to open a PR to backport this to 12.6 (apologies if I missed it)

**Comment by perrotta** [comment]
+1

**Comment by swagata87** [comment]
> @swagata87 if you can please to open a PR to backport this to 12.6 (apologies if I missed it)

I will do the backport now.
